### PR TITLE
Cherry-pick compaction enhancements and add range check function

### PR DESCRIPTION
**Pull request by tabokie**

### CODE REVIEW

File: db/db_impl/db_impl_compaction_flush.cc [hunk 1]
```diff
@@ -1108,79 +1107,102 @@ Status DBImpl::CompactRangeInternal(const CompactRangeOptions& options,
...
+        for (;;) {
+          // Always allow L0 -> L1 compaction
+          if (level > 0) {
+            if (cfd->ioptions()->level_compaction_dynamic_level_bytes) {
+              assert(final_output_level < cfd->ioptions()->num_levels);
+              if (final_output_level + 1 == cfd->ioptions()->num_levels) {
+                break;
+              }
+            } else {
+              // TODO(cbi): there is still a race condition here where
```
Comment (tonyxuqqi) [inline db/db_impl/db_impl_compaction_flush.cc:1145]: will this still cause the source region's trim to not clean all the data beyond its range?
What if we always set final_output_level to last level (6)

**Review by tonyxuqqi — APPROVED** [review]

File: db/db_impl/db_impl_compaction_flush.cc (hunk 1 above)
Comment (tabokie) [reply db/db_impl/db_impl_compaction_flush.cc:1145]: This is for dynamic_level_bytes=false, using 6 will move all files to L6 which makes write amp much larger for smaller dataset.

File: db/db_impl/db_impl_merge.cc [hunk 2]
```diff
@@ -57,6 +57,32 @@ Status DBImpl::ValidateForMerge(const MergeInstanceOptions& mopts,
   return Status::OK();
 }
 
+Status DBImpl::CheckInRange(const Slice* begin, const Slice* end) {
+  Status s;
+  if (begin == nullptr && end == nullptr) {
+    return s;
+  }
```
Comment (overvenus) [inline db/db_impl/db_impl_merge.cc:64]: What does `nullptr` mean? Should we return `Status::InvalidArgument`?

**Comment by overvenus** [comment]
Rest LGTM

File: db/db_impl/db_impl_merge.cc (hunk 2 above)
Comment (tabokie) [reply db/db_impl/db_impl_merge.cc:64]: It means infinity. It's a common usage of nullptr within RocksDB codebase.

**Review by overvenus — APPROVED** [review]

**Pull request merged**